### PR TITLE
fix: inconsistency in select and treeselect widgets behaviours

### LINKED ISSUE #7991: [Bug] The cursor is at the beginning of the TreeSelect widget even after selecting values
## Description
Expected Result:
It is observed that the insertion pointer is at the starting point in the tree-select even after selecting the value

Expected Result:
After selecting values the insertion pointer should be at the character ending point

### Steps to reproduce the behaviour:

1.Launch appsmith app in the release environment 
2.Drag and drop TreeSelect Widget 
3.Click onside the treeselect widget 
4.Choose any of the option and observe

Videolink:
[LOOM DEMO](https://www.loom.com/share/bfe3bf7b08b04d9c85165685667603c9)

### Important Details

- Version: [11.3]
- OS: [MacOS]
- Browser [chrome, safari]
- Environment [production, release]

### LINKED ISSUE #9430: [Bug]: Audio icon is not displayed in the entity explorer when the audio widget is dragged to the canvas
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

It is observed that when the audio widget is placed on the canvas the audio icon is not displayed in the entity explorer

[LOOM DEMO](https://www.loom.com/share/354f07dab1f74883a330f8aee31d45b6)

### Steps To Reproduce

1.Launch Appsmith in release/Prod Environment 
2.Drag and drop audio widget 
3.Check for the audio icon in the entity explorer and observe

### Environment

Production

### Version

Cloud

### LINKED ISSUE #9869: [Bug]: Treeselect, MultiSelect, MulitiTreeSelect widgets is set to required on initial state has red border
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

It is observed that when the multiselect, MulittreeSelect and treeselect widgets required is set to true and on the initial state red highlight is displayed 

[LOOM DEMO](https://www.loom.com/share/da2a7d8af9ad4e34bb6dc87ad4f6f518)

### Steps To Reproduce

1.launch appmsith app in release/Prod Environment
2.Drag and drop multiselect, MultiTreeSelect or Treeselect widget 
3.Make the required toggle enable 
4.Reload the page and observe

### Environment

Production

### Version

Cloud

### LINKED ISSUE #11073: [Bug]: Inconsistency in different Select widgets dropdown
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

The dropdown displays behavior is different in all types of select widgets. Since we should follow a similar functionality for entire select widgets

[LOOM DEMO](https://www.loom.com/share/c0617af60d89464e85a4d0bad439449b)

### Steps To Reproduce

1. Add select, multi select, tree select, multi tree select widget on canvas
2. Add  large text options into dropdown list to observe the inconsistency better
3. Observe on select widget: Displays entire text in dropdown which may cross canvas screen size
4. Observe on Multi select widget: Displays truncated text if text is large
5. Observe on Tree select & tree multi select: Displays truncated text with horizontal scroll

### Environment

Production

### Version

Cloud

### LINKED ISSUE #11076: [Bug]: Tree select widget behavior for long label names
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

[LOOM DEMO](https://www.loom.com/share/7e03fa19a4dc4099899905ad9d0c744e)

### Steps To Reproduce

1. Add Tree select widget on canvas
2. Add large text options in dropdown list
3. Observe that text been overlapped with dropdown icon

### Environment

Production

### Version

Cloud

### LINKED ISSUE #11614: Remove the "label" text inside label within select family widgets
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

By default when you drag an input, you don't see a label but you see a label for select widgets (select, multi-select, tree-select and multi-tree-select). Please remove the text "label" for consistency until we launch #7064 

### Why should this be worked on?

Top labels are not discoverable in other widgets and this top label also occupies a lot of space in the page. Users remove this and add text widgets as labels.

### PR DESCRIPTION
**Pull request by Tooluloope**
## Description

- Fix Uniformity among the select widget styles and behavior
- Fix invalid states in Select Widgets
- Fix Explorer icons for TreeSelect widgets
- Fix Dropdown text-overflow behavior in Select widgets
- Remove the default Label text from all Select widgets
- Remove Default focus from the first item in dropdown
- Make filter get focus once dropdown opens
- Added tooltip to SingleSelect and SingleTreeSelect
- Fixes issue with compact mode when select and label are inline, the dropdown tends to extend past the allowed width as shown below
Screenshot 2022-03-04 at 11 47 05

Fixes #11073 
Fixes #11076 
Fixes #7991
Fixes #9430
Fixes #9869
Fixes #11614

## Type of change
- Bugfix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- Manual testing

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: fix/inconsistent-select-widget-behaviours 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.82 **(-0.09)** | 37.16 **(-0.11)** | 35.21 **(-0.08)** | 56.14 **(-0.1)**
 :red_circle: | app/client/src/icons/WidgetIcons.tsx | 58.95 **(-0.19)** | 100 **(0)** | 15.22 **(-0.34)** | 58.95 **(-0.19)**
 :green_circle: | app/client/src/utils/WorkerUtil.ts | 89.76 **(0.78)** | 72.55 **(1.96)** | 100 **(0)** | 93.33 **(0.95)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**
 :green_circle: | app/client/src/widgets/MultiSelectTreeWidget/component/index.styled.tsx | 45.45 **(0.21)** | 25 **(0.76)** | 0 **(0)** | 50 **(0)**
 :red_circle: | app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx | 16.67 **(-8.82)** | 0 **(0)** | 0 **(0)** | 18.31 **(-9.35)**
 :red_circle: | app/client/src/widgets/MultiSelectTreeWidget/widget/index.tsx | 41.54 **(-2.01)** | 0 **(0)** | 42.86 **(0)** | 42.62 **(-2.21)**
 :green_circle: | app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx | 47.5 **(2.5)** | 31.25 **(3.12)** | 0 **(0)** | 52.94 **(1.42)**
 :red_circle: | app/client/src/widgets/MultiSelectWidgetV2/component/index.tsx | 12.04 **(-0.73)** | 0 **(0)** | 0 **(0)** | 13.54 **(-0.92)**
 :red_circle: | app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx | 80 **(-0.95)** | 50 **(-3.06)** | 70.59 **(0)** | 80 **(-1.01)**
 :green_circle: | app/client/src/widgets/SelectWidget/component/index.styled.tsx | 46.88 **(1.43)** | 23.33 **(1.45)** | 0 **(0)** | 50 **(0)**
 :red_circle: | app/client/src/widgets/SelectWidget/component/index.tsx | 21.74 **(-6.04)** | 0 **(0)** | 5.56 **(-1.11)** | 21.18 **(-6.09)**
 :green_circle: | app/client/src/widgets/SingleSelectTreeWidget/component/index.styled.tsx | 50 **(2.5)** | 25.81 **(0.81)** | 0 **(0)** | 52.94 **(1.42)**
 :red_circle: | app/client/src/widgets/SingleSelectTreeWidget/component/index.tsx | 16.88 **(-9.12)** | 0 **(0)** | 0 **(0)** | 18.57 **(-9.69)**
 :red_circle: | app/client/src/widgets/SingleSelectTreeWidget/widget/index.tsx | 49.12 **(-2.73)** | 0 **(0)** | 50 **(0)** | 50.94 **(-3.06)**</details>

### CODE REVIEW

**Comment by dilippitchika** [comment]
@Tooluloope Adding suggestions here
1. Can we remove the close icon in the filter bar when there's no text in the multi, multitree and tree select widgets. We should follow the behavior in the select widget.
2. Filter bar should also follow the same min width as the widget.
Screenshot 2022-03-07 at 12 26 16 PM
3. Can we show the label on hover in the selected mode when the label is truncated?
Screenshot 2022-03-07 at 12 27 40 PM
4. Fixes https://github.com/appsmithorg/appsmith/issues/9430 - this is not related to select and also not fixed in the DP. Either fix or remove it from the PR description.

**Comment by Tooluloope** [comment]
@dilippitchika Thanks, I'd work on all.
For the Forth, The icons are currently displayed
Screenshot 2022-03-07 at 08 16 56

**Comment by dilippitchika** [comment]
Hey @Tooluloope the issue is for the "Audio" widget and not "Audio recorder" widget

**Comment by Tooluloope** [comment]
@dilippitchika All issues are now fixed.

File: app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx [hunk 1]
```diff
@@ -127,23 +149,57 @@ function MultiTreeSelectComponent({
     }
     return document.querySelector(`.${CANVAS_CLASSNAME}`) as HTMLElement;
   }, []);
+  const onQueryChange = (event: ChangeEvent<HTMLInputElement>) => {
+    event.stopPropagation();
+    setFilter(event.target.value);
+  };
+
+  const memoDropDownWidth = useMemo(() => {
+    const parentWidth = width - WidgetContainerDiff;
```
Comment (sbalaji1192) [inline app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx:158]: move this two declaration after if condition

File: app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx [hunk 2]
```diff
@@ -127,23 +149,57 @@ function MultiTreeSelectComponent({
     }
     return document.querySelector(`.${CANVAS_CLASSNAME}`) as HTMLElement;
   }, []);
+  const onQueryChange = (event: ChangeEvent<HTMLInputElement>) => {
```
Comment (sbalaji1192) [inline app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx:152]: use useCallback

File: app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx [hunk 3]
```diff
@@ -127,23 +149,57 @@ function MultiTreeSelectComponent({
...
+            rightElement={clearButton as JSX.Element}
+            small
+            type="text"
+            value={filter}
+          />
+        ) : null}
+        <div className={`${loading ? Classes.SKELETON : ""}`}>{menu}</div>
+      </>
+    ),
+    [loading, isFilterable, filter],
```
Comment (sbalaji1192) [inline app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx:189]: add `onQueryChange` as well

File: app/client/src/widgets/MultiSelectTreeWidget/widget/index.tsx [hunk 4]
```diff
@@ -412,6 +415,9 @@ class MultiSelectTreeWidget extends BaseWidget<
         type: EventType.ON_OPTION_CHANGE,
       },
     });
+    if (!this.props.isDirty) {
```
Comment (sbalaji1192) [inline app/client/src/widgets/MultiSelectTreeWidget/widget/index.tsx:418]: Have you checked with @wmdev0808 on introducing this property?

File: app/client/src/widgets/MultiSelectWidgetV2/component/index.tsx [hunk 5]
```diff
@@ -89,17 +89,19 @@ function MultiSelectComponent({
...
+    () =>
+      filter ? (
+        <Button
+          disabled={disabled}
+          icon="cross"
+          minimal
+          onClick={() => setFilter("")}
+        />
+      ) : null,
+    [filter],
```
Comment (sbalaji1192) [inline app/client/src/widgets/MultiSelectWidgetV2/component/index.tsx:104]: add `disabled` to deps

File: app/client/src/widgets/MultiSelectWidgetV2/component/index.tsx [hunk 6]
```diff
@@ -184,6 +186,17 @@ function MultiSelectComponent({
     },
     serverSideFiltering ? [options] : [filter, options],
   );
+  const memoDropDownWidth = useMemo(() => {
+    const parentWidth = width - WidgetContainerDiff;
```
Comment (sbalaji1192) [inline app/client/src/widgets/MultiSelectWidgetV2/component/index.tsx:190]: move two const after if condition.

File: app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx [hunk 7]
```diff
@@ -509,7 +510,7 @@ export interface MultiSelectWidgetProps extends WidgetProps {
   options?: DropdownOption[];
   onOptionChange: string;
   onFilterChange: string;
-  defaultOptionValue: string | string[];
+  defaultOptionValue: string | string[] | { value: string; label: string }[];
```
Comment (sbalaji1192) [inline app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx:513]: lets create a type for this `{ value: string; label: string }`

File: app/client/src/widgets/SelectWidget/component/index.tsx [hunk 8]
```diff
@@ -98,15 +119,21 @@ class SelectComponent extends React.Component<
...
 
+    // Check if text overflows
+    const tooltipText =
+      this.spanRef.current?.parentElement &&
+      (this.spanRef.current.parentElement.offsetHeight <
+        this.spanRef.current.parentElement.scrollHeight ||
+        this.spanRef.current.parentElement.offsetWidth <
+          this.spanRef.current.parentElement.scrollWidth)
+        ? value
+        : undefined;
```
Comment (sbalaji1192) [inline app/client/src/widgets/SelectWidget/component/index.tsx:130]: should this be empty string ?

File: app/client/src/widgets/SelectWidget/widget/index.tsx [hunk 9]
```diff
@@ -421,6 +415,9 @@ class SelectWidget extends BaseWidget<SelectWidgetProps, WidgetState> {
         },
       });
     }
+    if (!this.props.isDirty) {
```
Comment (sbalaji1192) [inline app/client/src/widgets/SelectWidget/widget/index.tsx:418]: Just curious to know whats the significance of moving this here from top?

File: app/client/src/widgets/SingleSelectTreeWidget/component/index.tsx [hunk 10]
```diff
@@ -125,21 +133,69 @@ function SingleSelectTreeComponent({
...
+    () =>
+      filter ? (
+        <Button
+          disabled={disabled}
+          icon="cross"
+          minimal
+          onClick={() => setFilter("")}
+        />
+      ) : null,
+    [filter],
```
Comment (sbalaji1192) [inline app/client/src/widgets/SingleSelectTreeWidget/component/index.tsx:147]: add `disabled`

File: app/client/src/widgets/SingleSelectTreeWidget/component/index.tsx [hunk 11]
```diff
@@ -125,21 +133,69 @@ function SingleSelectTreeComponent({
...
+  );
+  const onQueryChange = (event: ChangeEvent<HTMLInputElement>) => {
+    event.stopPropagation();
+    setFilter(event.target.value);
+  };
+
+  const memoDropDownWidth = useMemo(() => {
+    const parentWidth = width - WidgetContainerDiff;
```
Comment (sbalaji1192) [inline app/client/src/widgets/SingleSelectTreeWidget/component/index.tsx:155]: move this declarations below if condition

File: app/client/src/widgets/SingleSelectTreeWidget/component/index.tsx [hunk 12]
```diff
@@ -125,21 +133,69 @@ function SingleSelectTreeComponent({
...
+            rightElement={clearButton as JSX.Element}
+            small
+            type="text"
+            value={filter}
+          />
+        ) : null}
+        <div className={`${loading ? Classes.SKELETON : ""}`}>{menu}</div>
+      </>
+    ),
+    [loading, isFilterable, filter],
```
Comment (sbalaji1192) [inline app/client/src/widgets/SingleSelectTreeWidget/component/index.tsx:187]: add `onQueryChange`

File: app/client/src/widgets/SingleSelectTreeWidget/component/index.styled.tsx [hunk 13]
```diff
@@ -263,6 +261,54 @@ ${({ dropDownWidth, id, parentWidth }) => `
     box-shadow: none;
     outline: none !important;
   }
+  & .${Classes.INPUT_GROUP} {
```
Comment (sbalaji1192) [inline app/client/src/widgets/SingleSelectTreeWidget/component/index.styled.tsx:264]: L264 to L311 is copy of L299 to L346 from `MultiSelectTreeWidget/component/index.styled.tsx`. Shall we create a common function that returns these styles.

**Review by sbalaji1192 — COMMENTED** [review]
Avoid using !important, across the changes

File: app/client/src/widgets/MultiSelectTreeWidget/widget/index.tsx (hunk 4 above)
Comment (Tooluloope) [reply app/client/src/widgets/MultiSelectTreeWidget/widget/index.tsx:418]: Yeah, Checked his PR already

File: app/client/src/widgets/SelectWidget/widget/index.tsx (hunk 9 above)
Comment (Tooluloope) [reply app/client/src/widgets/SelectWidget/widget/index.tsx:418]: Oh, I wanted to put it inside the `isChanged` condition.
So before I proceed with this, I need to clarify something.

Would the Select be dirty, if all the user did was to Select an option that was already selected, in this case, the value didn't change and `isChanged` is still false?

Comment (sbalaji1192) [reply app/client/src/widgets/SelectWidget/widget/index.tsx:418]: Ideally, it should become dirty only when the value changes.  @dilippitchika correct me if I'm wrong

Comment (dilippitchika) [reply app/client/src/widgets/SelectWidget/widget/index.tsx]: Agreed with @sbalaji1192

**Review by sbalaji1192 — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@Tooluloope Tested issue #7991 and it is working as expected now

**Comment by YogeshJayaseelan** [comment]
@Tooluloope Tested issue #9869 and it is working as expected

**Comment by shwetha-ramesh** [comment]
@Tooluloope tested #9430. DP is working as expected
image

**Comment by YogeshJayaseelan** [comment]
@Tooluloope Tested issue #11614 - Now for all the select family by default 'label' text is not displayed unless the user configures label text. Working as expected

**Comment by shwetha-ramesh** [comment]
@Tooluloope tested #11073. Found 2 issues.
1. No ellipsis found in Select widget when URL is an option
image

2. For select widget in deploy mode the first option remains highlighted with green
image

**Review by dilippitchika — APPROVED** [review]

**Comment by shwetha-ramesh** [comment]
@Tooluloope re-tested #11073 DP after fix. Working as expected. Moving to Done.

**Pull request merged**